webpay/pay/tasks.py

```diff
@@ -275,13 +280,67 @@ def start_pay(transaction_uuid, notes, user_uuid, provider_names, **kw):
             'status': constants.STATUS_PENDING
         })
     except Exception, exc:
+        etype, val, tb = sys.exc_info()
+        pay_error_handler(
+            error_type=etype,
+            provider_helper=provider_helper,
+            source=source,
+            uuid=transaction_uuid,
+        )
         log.exception('while configuring payment for transaction {t}: '
                       '{exc.__class__.__name__}: {exc}'
                       .format(t=transaction_uuid, exc=exc))
-        etype, val, tb = sys.exc_info()
         raise exc, None, tb
 
 
+def pay_error_handler(
+        error_type,
+        provider_helper,
+        source,
+        uuid):
+    """
+    Record a transaction error into solitude. At this point the transaction
+    may, or may not exist in solitude.
+
+    Unfortunately many things (buyer, seller, seller_product) are buried inside
+    start_transaction, which might need its own wrapper.
+    """
+    pk = None
+    api = client.slumber.generic.transaction
+    try:
+        pk = api.get_object_or_404(uuid=uuid)
+    except ObjectDoesNotExist:
+        pass
+
+    # If the provider_helper is None, then no provider was found.
+    provider = None
+    # Convert the helper into a solitude constant.
+    if provider_helper:
+        provider = constants.PROVIDERS[provider_helper.name]
+
+    if pk:
+        # For a transaction to exist, everything up to creating the
+        # transaction in the payment provider has worked and something post
+        # that (eg in getting and patching the transaction) has failed.
+        #
+        # This should be pretty unusual.
+        log.info('Updating transaction as error for: {0}'.format(uuid))
+        api(pk).patch({
+            'status_reason': 'POST_CREATION_FAILURE',
```

Comment:
why can't ``status_reason`` also be ``getattr(error_type, 'solitude', 'UNKNOWN')`` ?